### PR TITLE
Fix property ordering for delayed values

### PR DESCRIPTION
**Pull request by fromcelticpark**
Resolves object property insertion order issue described here: https://github.com/facebook/prepack/issues/730

### CODE REVIEW

File: src/serializer/ResidualHeapSerializer.js [hunk 1]
```diff
@@ -1101,6 +1101,13 @@ export class ResidualHeapSerializer {
                   mightHaveBeenDeleted
                 );
               });
+
+              // Although the property needs to be delayed, we still want to emit "undefined"
+              // as part of the object literal to ensure a consistent property ordering.
+              if (!mightHaveBeenDeleted) {
```
Comment (fromcelticpark) [inline src/serializer/ResidualHeapSerializer.js]: I'm not really sure, that `!mightHaveBeenDeleted` is the right condition here to check against.
But without this check tests were failing with the following error:

```
test/serializer/abstract/PutValue11.js
Output mismatch!
original code
var c = global.__abstract ? __abstract("boolean", "false") : false;
a = {};
if (c) a.f = a;

inspect = function() { return 'f' in a; }

output of inspect() on original code
false
generated code in iteration 0
var c;
(function () {
  var __empty = {};

  var _$0_unique27277 = this;

  function _3_unique27277() {
    return 'f' in _$0_unique27277.a;
  }

  c = void 0;
  var _0_unique27277 = false;
  c = _0_unique27277;
  var _1_unique27277 = {
    f: void 0
  };

  var _2_unique27277 = _0_unique27277 ? _1_unique27277 : __empty;

  if (_2_unique27277 !== __empty) _1_unique27277.f = _2_unique27277;
  a = _1_unique27277;
  inspect = _3_unique27277;
}).call(this);
output of inspect() on last generated code iteration
true
```

Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Your change certainly improved things into the right direction, and checking for `!mightHaveBeenDeleted` makes sense.

Even better, but that could come in a different pull request, would be if you always emit the `undefined` entry here, but then issue a (conditional) `delete` statement afterwards if the entry got deleted. In `_assignProperty`, the assignment currently happens conditionally, and there an `else` branch could be generated that issue the deletion, assuming that your new kicks in that adds a dummy value to begin with (`_assignProperty` is called from a few more places.)

Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: I think this test case brings out the issue:
```js
function() {
    let x = global.__abstract ? __abstract("boolean", "false") : false;
    let o = { a: undefined, b: {} };
    o.a = o;
    if (x) delete o.a;

    global.inspect = function() {
      for (let name in o) { return name; }
    };
})();
```

Can you either address it as described above, or create a new issue to track the remaining work?

Comment (fromcelticpark) [inline src/serializer/ResidualHeapSerializer.js]: Yeah, I will look into it in the current pull request.

**Pull request closed without merging**